mcan/src/lib.rs

```diff
@@ -1,6 +1,263 @@
+#![warn(missing_docs)]
+//! # MCAN
+//!
+//! ## Overview
+//! This crate provides a platform-agnostic CAN HAL.
+//!
+//! It provides the following features:
+//!
+//! - classical CAN and CAN FD with bitrate switching support
+//! - safe Message RAM layouting system via [`SharedMemory`] that prevents
+//!   misconfiguration through compile-time checks
+//! - modular interrupt handling abstractions that enable lock-less usage of HW
+//!   interrupt lines
+//! - message transmission using dedicated buffers, FIFO and priority queue
+//! - message transmission cancellation
+//! - message reception using dedicated buffers and two FIFOs
+//! - filter settings
+//!
+//! MCAN is embedded in the MCU like all other peripherals. The interface
+//! between them includes two clock signal lines, two HW interrupt lines, a
+//! single memory-mapped HW register and a dedicated, shared RAM memory region
+//! (referred to as Message RAM) that both CPU and MCAN can access and share
+//! information through.
+//!
+//! For the MCAN abstractions to be considered operational, this interface has
+//! to be properly configured. The latter is assured through the safety
+//! requirements of [`mcan_core`] traits which platform-specific HALs are
+//! expected to implement.
+//!
+//! In order to use MCAN, one has to instantiate [`CanConfigurable`] and
+//! [`finalize`] it. Its constructor requires an instance of an
+//! [`Dependencies`] implementing struct and holds onto it until it's
+//! [`freed`]. Safety requirements of the `Dependencies` trait
+//! guarantee a correct state of MCAN interfaces during its operation.
+//!
+//! TODO: We do not have `CanConfigurable::free`??
+//!
+//! ## Message RAM Configuration
+//!
+//! All the platform specific details are covered by
+//! `Dependencies` **apart from** the Message RAM configuration.
+//! This is because it is hard to enclose all the safety requirements for the
+//! shared message RAM on a platform-specific HAL level.
+//!
+//! The MCAN uses 16-bit addressing internally. It means that all higher bytes
+//! of the addressing has to be configured externally to the MCAN HAL.
+//! [`Dependencies::eligible_message_ram_start`] implemented by
+//! platform-specific HAL provides a way to `mcan` to verify if the memory
+//! region provided by a user is sound; yet it is up to the user to put it in a
+//! valid, accessible to MCAN, RAM memory region.
+//!
+//! One can configure the Message RAM as follows
+//! - specify a custom MEMORY entry in a linker script mapped to the valid RAM
+//!   memory region
+//! - introduce a custom, `.bss` like (NOLOAD property), section - eg. `.can`
+//! - map a new custom input section to a new custom MEMORY entry
+//! - use the `#[link_section]` attribute in a code to link a static variable to
+//!   this memory region
+//!
+//! Example of a linker script
+//! ```text
+//! MEMORY
+//! {
+//!   FLASH : ORIGIN = 0x400000, LENGTH = 2M
+//!   CAN : ORIGIN = 0x20400000, LENGTH = 64K
```

Comment:
One MCAN can use up to 4352 32-bit words (17408 bytes) of RAM. This 64 KiB region is enough for 3 full MCAN peripherals, or 4 in a slightly less extreme configuration. I think the example should show 17 K.

![image](https://user-images.githubusercontent.com/108275985/207820977-79978250-3bd7-4ace-945a-f6b1d229680c.png)

Reply:
I think this is fine

Reply:
I guess this is OK since space-constrained users will want to tailor the size anyway.